### PR TITLE
Implement 'GetSystemTimeAsFileTime' shim for Windows

### LINKED ISSUE #997: Implement `GetSystemTimeAsFileTime` shim for Windows
`SystemTime::now` is [implemented](https://github.com/rust-lang/rust/blob/e413dc36a83a5aad3ab6270373000693a917e92b/src/libstd/sys/windows/time.rs#L82) in windows using the `GetSystemTimeAsFileTime` foreign function, which currently doesn't have a shim in miri.

**Some instructions:**

Implementing the shim for `GetSystemTimeAsFileTime` its pretty similar to the linux (`clock_gettime`) and macOS (`gettimeofday`) shims (those can already be found in `src/shims/time.rs`). Its functionality is described in detail [here](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsystemtimeasfiletime).

The main difference is that this shim only has one argument: a pointer to a [`FILETIME` structure](https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-filetime), which has two integer fields `dwLowDateTime` and `dwHighDateTime`. These fields correspond to the first and last 32 bits of the 64 bits number counting the 100-nanosecond intervals (so its basically an `u64` represented as two `u32`).

It is also important to take into account that windows counts time since January 1, 1601 instead of January 1, 1970.

The testcase for this issue is:
```rust
use std::time::SystemTime;

fn main() {
   let _now = SystemTime::now();
}
```
Which can be found in `miri/tests/run-pass/time.rs`.

### PR DESCRIPTION
**Pull request by JOE1994**
Implement `GetSystemTimeAsFileTime` shim for Windows.

Closes #997

### CODE REVIEW

**Comment by RalfJung** [comment]
Instead of splitting the test, please use `#[cfg]` to only test the supported things on each platform.

**Comment by RalfJung** [comment]
See https://github.com/rust-lang/miri/pull/1287 for how to use `cfg` to only run parts of a test on some platform.

File: tests/run-pass/time.rs [hunk 1]
```diff
@@ -4,15 +4,19 @@
 use std::time::{SystemTime, Instant};
 
 fn main() {
+
```
Comment (RalfJung) [inline tests/run-pass/time.rs]: What's with the newline here?
Also you need to remove `ignore-windows` if this is supposed to test anything. :)

File: tests/run-pass/time.rs [hunk 2]
```diff
@@ -4,15 +4,19 @@
...
     for _ in 0..10 { drop(vec![42]); }
     let now2 = SystemTime::now();
     assert!(now2 > now1);
 
-    let now1 = Instant::now();
-    // Do some work to make time pass.
-    for _ in 0..10 { drop(vec![42]); }
-    let now2 = Instant::now();
-    assert!(now2 > now1);
+    // Shims not yet implemented on Windows
```
Comment (RalfJung) [inline tests/run-pass/time.rs]:
```suggestion
    // `Instant` shims not yet implemented on Windows
```

File: src/shims/time.rs [hunk 3]
```diff
@@ -85,6 +86,33 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+    #[allow(non_snake_case)]
+    fn GetSystemTimeAsFileTime(&mut self, LPFILETIME_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        this.assert_target_os("windows", "GetSystemTimeAsFileTime");
+        this.check_no_isolation("GetSystemTimeAsFileTime")?;
+
+        // A Windows tick is 100 nanoseconds. Windows epoch 1601-01-01T00:00:00Z
+        // is 11644473600 seconds before Unix epoch 1970-01-01T00:00:00Z.
+        let duration = system_time_to_duration(&SystemTime::now())?
```
Comment (RalfJung) [inline src/shims/time.rs]: Please reuse the constants from libstd:
```rust
const NANOS_PER_SEC: u64 = 1_000_000_000;
const INTERVALS_PER_SEC: u64 = NANOS_PER_SEC / 100;
const INTERVALS_TO_UNIX_EPOCH: u64 = 11_644_473_600 * INTERVALS_PER_SEC;
```

File: src/shims/time.rs [hunk 4]
```diff
@@ -85,6 +86,33 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+            .checked_add(Duration::from_secs(11644473600))
+            .unwrap();
+        let windows_nanoseconds_per_tick = 100;
+        let duration_ticks = u64::try_from(duration.as_nanos() / windows_nanoseconds_per_tick)
+            .map_err(|_| err_unsup_format!("programs running longer than 2^64 nanoseconds are not supported"))?;
```
Comment (RalfJung) [inline src/shims/time.rs]:
```suggestion
            .map_err(|_| err_unsup_format!("programs running longer than 2^64 ticks are not supported"))?;
```

File: src/shims/time.rs [hunk 5]
```diff
@@ -85,6 +86,33 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+
+        let DWORD_tylayout = this.windows_ty_layout("DWORD")?;
```
Comment (RalfJung) [inline src/shims/time.rs]: We are encoding DWORD = u32 everywhwere, so you might as well also do that here.
```suggestion
        let DWORD_tylayout = this.layout_of(this.tcx.types.u32)?;
```

File: src/shims/time.rs [hunk 6]
```diff
@@ -85,6 +86,33 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+        let dwLowDateTime = u32::try_from(duration_ticks & 0x00000000FFFFFFFF).unwrap();
```
Comment (RalfJung) [inline src/shims/time.rs]: In partciular you are hard-coding the `u32` here anyway :)

File: src/shims/time.rs [hunk 7]
```diff
@@ -1,9 +1,10 @@
 use std::time::{Duration, SystemTime, Instant};
 use std::convert::TryFrom;
+use std::ops::Shr;
```
Comment (RalfJung) [inline src/shims/time.rs]: Why not use the `>>` syntax?

File: tests/run-pass/time.rs [hunk 8]
```diff
@@ -4,15 +4,19 @@
...
+    #[cfg(not(windows))] {
```
Comment (RalfJung) [inline tests/run-pass/time.rs:15]: Yes, looking good. :)
This is also fine for the PR; if you plan to implement `Instant::now()`, please make that a separate PR.

**Comment by JOE1994** [comment]
I squashed the fixes to a single commit.

File: src/shims/time.rs [hunk 9]
```diff
@@ -85,6 +87,36 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+        this.assert_target_os("windows", "GetSystemTimeAsFileTime");
+        this.check_no_isolation("GetSystemTimeAsFileTime")?;
+
+        let NANOS_PER_SEC = this.eval_windows_u64("time", "NANOS_PER_SEC")?;
+        let INTERVALS_PER_SEC = this.eval_windows_u64("time", "INTERVALS_PER_SEC")?;
+        let INTERVALS_TO_UNIX_EPOCH = this.eval_windows_u64("time", "INTERVALS_TO_UNIX_EPOCH")?;
+        let NANOS_PER_INTERVAL = NANOS_PER_SEC / INTERVALS_PER_SEC;
+        let SECONDS_TO_UNIX_EPOCH = INTERVALS_TO_UNIX_EPOCH / INTERVALS_PER_SEC;
+
+        let duration = system_time_to_duration(&SystemTime::now())?
```
Comment (retep998) [inline src/shims/time.rs:103]: This will fail if `rustc` is running on a Windows system with its clock configured to a time before the unix epoch.

Comment (RalfJung) [reply src/shims/time.rs:103]: That is already true however for e.g. Unix targets running on Windows hosts. The error in that case will be "times before the Unix epoch are not supported", which is reasonably accurate. And the usecase is sufficiently niche that I am okay with not supporting this initially.

(Patches/bugreports still welcome, of course.)

Comment (retep998) [reply src/shims/time.rs:103]: For unix targets on Windows hosts, I would personally prefer saturation as a solution instead.

Comment (RalfJung) [reply src/shims/time.rs:103]: Don't many targets support negative times though? Linux at least does, AFAIK. So some times before the epoch could still be supported, we just don't, for now -- because why would we spend any effort on systems with their clock before 1970?^^

Comment (RalfJung) [reply src/shims/time.rs:103]: Also I'd prefer to fail loudly over silently doing the wrong thing (like saturation).

File: src/shims/time.rs [hunk 10]
```diff
@@ -85,6 +87,36 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+    #[allow(non_snake_case)]
+    fn GetSystemTimeAsFileTime(&mut self, LPFILETIME_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        this.assert_target_os("windows", "GetSystemTimeAsFileTime");
+        this.check_no_isolation("GetSystemTimeAsFileTime")?;
+
+        let NANOS_PER_SEC = this.eval_windows_u64("time", "NANOS_PER_SEC")?;
+        let INTERVALS_PER_SEC = this.eval_windows_u64("time", "INTERVALS_PER_SEC")?;
+        let INTERVALS_TO_UNIX_EPOCH = this.eval_windows_u64("time", "INTERVALS_TO_UNIX_EPOCH")?;
```
Comment (RalfJung) [inline src/shims/time.rs:99]: Oh, I was actually thinking of copy-pasting those constants, but I guess this works as well.^^

**Comment by RalfJung** [comment]
This looks good, thanks. :)
@bors r+
